File: 1197-parsing-a-boolean-expression/1197-parsing-a-boolean-expression.java
class Solution {
    public boolean parseBoolExpr(String expression) {
        Stack<Character> st = new Stack<>();
        char temp = ' ';

        for (char ch : expression.toCharArray()) {
            if (ch == '(' || ch == ',') continue; // Skip '(' and ','

            // Push 't', 'f', '!', '&', '|' into the stack
            if (ch == 't' || ch == 'f' || ch == '!' || ch == '&' || ch == '|') {
                st.push(ch);
            } 
            // Process when encountering a closing parenthesis ')'
            else if (ch == ')') {
                boolean hasTrue = false;
                boolean hasFalse = false;

                // Pop boolean values inside the parentheses until reaching the operator
                while (!st.isEmpty() && (st.peek() == 't' || st.peek() == 'f')) {
                    char val = st.pop();
                    if (val == 't') hasTrue = true;
                    if (val == 'f') hasFalse = true;
                }

                // Pop the operator ('!', '&', or '|') if stack is not empty
                if (!st.isEmpty()) {
                    char operator = st.pop();

                    // Evaluate the result based on the operator
                    if (operator == '|') {
                        temp = hasTrue ? 't' : 'f';  // OR operation
                    } else if (operator == '&') {
                        temp = hasFalse ? 'f' : 't';  // AND operation
                    } else if (operator == '!') {
                        temp = hasTrue ? 'f' : 't';  // NOT operation
                    }

                    // Push the evaluated result back onto the stack
                    st.push(temp);
                }
            }
        }

        // The final result is at the top of the stack, if the stack is not empty
        return !st.isEmpty() && st.peek() == 't';
    }
}

File: 1197-parsing-a-boolean-expression/1197-parsing-a-boolean-expression.py
class Solution:
    def parseBoolExpr(self, expression: str) -> bool:
        st = []
        temp = ' '

        for ch in expression:
            if ch == '(' or ch == ',':
                continue  # Skip '(' and ','

            if ch in ['t', 'f', '!', '&', '|']:
                st.append(ch)
            elif ch == ')':
                has_true = False
                has_false = False

                # Process until reaching the operator
                while st and st[-1] in ['t', 'f']:
                    val = st.pop()
                    if val == 't':
                        has_true = True
                    if val == 'f':
                        has_false = True

                # Pop the operator if stack isn't empty
                if st:
                    operator = st.pop()

                    # Evaluate the result based on the operator
                    if operator == '|':
                        temp = 't' if has_true else 'f'  # OR operation
                    elif operator == '&':
                        temp = 'f' if has_false else 't'  # AND operation
                    elif operator == '!':
                        temp = 'f' if has_true else 't'  # NOT operation

                    # Push the result back to stack
                    st.append(temp)

        # The final result is the last element in the stack
        return st and st[-1] == 't'
